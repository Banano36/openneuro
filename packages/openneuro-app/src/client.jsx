/**
 * Browser client entrypoint - see server.tsx for SSR entrypoint
 */
import './scripts/utils/global-polyfill'
import './scripts/apm.js'
import { ApolloProvider, InMemoryCache } from '@apollo/client'
import { createClient } from '@openneuro/client'
import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter, Route } from 'react-router-dom'
import App from './scripts/app'
import Index from './scripts/index'
import analyticsWrapper from './scripts/utils/analytics'
import { version } from './lerna.json'
import { config } from './scripts/config'
import * as gtag from './scripts/utils/gtag'
import { relayStylePagination } from '@apollo/client/utilities'
import redesignStyles from '@openneuro/components/page/page.scss'

gtag.initialize(config.analytics.trackingIds)

ReactDOM.render(
  <App>
    <ApolloProvider
      client={createClient(`${config.url}/crn/graphql`, {
        clientVersion: version,
        enableWebsocket: false,
        cache: new InMemoryCache({
          typePolicies: {
            Query: {
              fields: {
                advancedSearch: relayStylePagination(),
              },
            },
          },
        }),
      })}>
      <BrowserRouter>
        <Route component={analyticsWrapper(Index)} />
      </BrowserRouter>
    </ApolloProvider>
  </App>,
  document.getElementById('main'),
)
