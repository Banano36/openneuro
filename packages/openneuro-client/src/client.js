import 'cross-fetch/polyfill'
import { ApolloClient } from 'apollo-client'
import { setContext } from 'apollo-link-context'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { createUploadLink } from 'apollo-upload-client'
import { ApolloLink, split, Observable } from 'apollo-link'
import { WebSocketLink } from 'apollo-link-ws'
import { getMainDefinition } from 'apollo-utilities'
import FormData from 'form-data'
import semver from 'semver'
import * as files from './files'
import * as datasets from './datasets'
import * as snapshots from './snapshots'
import * as users from './users'

const cache = new InMemoryCache({
  freezeResults: true,
})

/**
 * Setup a client for working with the OpenNeuro API
 *
 * @param {string} uri
 */
const createClient = (uri, getAuthorization, fetch, clientVersion) => {
  const link = createLink(uri, getAuthorization, fetch, clientVersion)
  return new ApolloClient({ uri, link, cache, connectToDevTools: true })
}

const authLink = getAuthorization =>
  setContext((_, { headers }) => {
    // Passthrough any headers but add in authorization if set
    const token = getAuthorization()
    let cookie = {}
    if (token) {
      const tokenString = `accessToken=${token}`
      cookie = {
        cookie: tokenString,
      }
    }
    return {
      headers: Object.assign(cookie, headers),
    }
  })

const wsLink = uri => {
  const root = uri.replace('http', 'ws').replace('/crn', '')
  const subscriptions = '-subscriptions'
  const link = root + subscriptions
  return new WebSocketLink({
    uri: link,
    options: {
      reconnect: true,
    },
  })
}

const middlewareAuthLink = (uri, getAuthorization, fetch) => {
  // We have to setup authLink to inject credentials here
  const httpUploadLink = createUploadLink({
    uri,
    fetch: fetch === null ? undefined : fetch,
    serverFormData: FormData,
    credentials: 'same-origin',
  })
  if (getAuthorization) {
    return authLink(getAuthorization).concat(httpUploadLink)
  } else {
    return httpUploadLink
  }
}

const parse = version => [semver.major(version), semver.minor(version)]
const checkVersions = (serverVersion, clientVersion) => {
  if ([serverVersion, clientVersion].every(semver.valid)) {
    const [serverMajor, serverMinor] = parse(serverVersion)
    const [clientMajor, clientMinor] = parse(clientVersion)
    if (serverMajor > clientMajor || serverMinor > clientMinor) {
      console.warn(
        `Your openNeuro client is out of date (v${clientVersion}). We strongly recommend you update to the latest version (v${serverVersion}) for an optimal experience.`,
      )
    } else if (
      serverMajor < clientMajor ||
      (serverMajor === clientMajor && serverMinor < clientMinor)
    ) {
      // panic, then
      console.warn(
        'Your openNeuro client is out of date. We strongly recommend you update to the most recent version for an optimal experience.',
      )
    }
  }
}

const compareVersionsLink = clientVersion =>
  new ApolloLink(
    (operation, forward) =>
      new Observable(observer =>
        forward(operation).subscribe({
          next: result => {
            const serverVersion = result.extensions.openneuro.version
            // alert user if major/minor versions are not in sync
            checkVersions(serverVersion, clientVersion)
            observer.next(result)
          },
          error: console.error,
          complete: observer.complete.bind(observer),
        }),
      ),
  )

const createLink = (uri, getAuthorization, fetch, clientVersion) => {
  // We have to setup authLink to inject credentials here

  // server-side link
  let link = middlewareAuthLink(uri, getAuthorization, fetch)
  let ws

  try {
    // browser-side link
    ws = process.browser ? wsLink(uri) : null
  } catch (e) {
    // Don't die when websocket setup fails
  }
  if (ws) {
    link = split(
      ({ query }) => {
        const { kind, operation } = getMainDefinition(query)
        return kind === 'OperationDefinition' && operation === 'subscription'
      },
      ws,
      middlewareAuthLink(uri, getAuthorization, fetch),
    )
  }

  return ApolloLink.from([compareVersionsLink(clientVersion), link])
}

export { files, datasets, snapshots, users }
export default createClient
