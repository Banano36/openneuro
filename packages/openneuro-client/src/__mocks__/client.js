import { ApolloClient, InMemoryCache } from '@apollo/client'
import { SchemaLink } from '@apollo/client/link/schema'
import { addMocksToSchema } from '@graphql-tools/mock'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { typeDefs } from '@openneuro/server/src/graphql/schema'
import resolvers from '@openneuro/server/src/graphql/resolvers/index.js'

/**
 * Generate sequential dataset ids for tests
 * @param {number} n Starting string value
 */
function* idGenerator(n = 0) {
  while (true) {
    const idStr = String(++n).padStart(6, '0')
    yield `ds${idStr}`
  }
}

export const datasetIds = idGenerator()
export const testDsId = datasetIds.next().value
export const testTime = new Date()

// Workaround to restore this test, it should be refactored with openneuro-client updates
const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
})

addMocksToSchema({
  schema,
  mocks: {
    Dataset: (root, { id }) => ({
      id: () => (id ? id : datasetIds.next().value),
      created: testTime,
      modified: testTime,
    }),
    DateTime: () => {
      return testTime
    },
    BigInt: () => {
      return 8589934592
    },
  },
})

export const createClient = uri => {
  const cache = new InMemoryCache()
  const link = new SchemaLink({
    schema,
    context: {
      user: '1234',
      userInfo: {
        id: '1234',
      },
    },
  })
  // @ts-expect-error
  return new ApolloClient({ uri, link, cache })
}
