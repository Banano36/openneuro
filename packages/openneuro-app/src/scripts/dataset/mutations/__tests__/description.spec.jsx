import React from "react"
import { render } from "@testing-library/react"
import { MockedProvider } from "@apollo/client/testing"
import UpdateDescription, {
  mergeFieldValue,
  UPDATE_DESCRIPTION,
  UPDATE_DESCRIPTION_LIST,
} from "../description.jsx"

const mockMutation = vi.fn()
vi.mock("@apollo/client/react/components", () => ({
  ...vi.importActual("@apollo/client/react/components"),
  Mutation: (props) => {
    mockMutation(props)
    return <>Mutation mock</>
  },
}))

describe("UpdateDescription mutation", () => {
  it("renders with common props", () => {
    const { asFragment } = render(
      <UpdateDescription
        datasetId="ds001"
        field="Name"
        value="New Name"
        done={vi.fn()}
      />,
      { wrapper: MockedProvider },
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("uses the scalar mutation for scalar values", () => {
    render(
      <UpdateDescription
        datasetId="ds001"
        field="Name"
        value="New Name"
        done={vi.fn()}
      />,
      { wrapper: MockedProvider },
    )
    expect(mockMutation).toHaveBeenCalledWith(
      expect.objectContaining({ mutation: UPDATE_DESCRIPTION }),
    )
  })
  it("uses the list mutation for array values", () => {
    render(
      <UpdateDescription
        datasetId="ds001"
        field="Authors"
        value={["John Doe", "Jane Doe"]}
        done={vi.fn()}
      />,
      { wrapper: MockedProvider },
    )
    expect(mockMutation).toHaveBeenCalledWith(
      expect.objectContaining({ mutation: UPDATE_DESCRIPTION_LIST }),
    )
  })
  describe("mergeFieldValue()", () => {
    it("merges in scalar fields", () => {
      expect(
        mergeFieldValue(
          "ds001",
          {
            __typename: "Draft",
            id: "1234",
            created: "1999",
            description: {
              Name: "Old Name",
              BIDSVersion: "1.2",
              License: "AGPL3",
            },
          },
          { Name: "New Name", BIDSVersion: "1.2", License: "AGPL3" },
          null,
        ),
      ).toEqual({
        __typename: "Dataset",
        draft: {
          __typename: "Draft",
          created: "1999",
          description: {
            BIDSVersion: "1.2",
            License: "AGPL3",
            Name: "New Name",
          },
          id: "1234",
        },
        id: "ds001",
      })
    })
    it("merges in array fields", () => {
      expect(
        mergeFieldValue(
          "ds001",
          {
            __typename: "Draft",
            id: "1234",
            created: "1999",
            description: {
              Name: "Old Name",
              BIDSVersion: "1.2",
              License: "AGPL3",
              Authors: [],
            },
          },
          null,
          {
            Name: "New Name",
            BIDSVersion: "1.2",
            License: "AGPL3",
            Authors: ["One", "Two"],
          },
        ),
      ).toEqual({
        __typename: "Dataset",
        draft: {
          __typename: "Draft",
          created: "1999",
          description: {
            Authors: ["One", "Two"],
            BIDSVersion: "1.2",
            License: "AGPL3",
            Name: "New Name",
          },
          id: "1234",
        },
        id: "ds001",
      })
    })
  })
})
