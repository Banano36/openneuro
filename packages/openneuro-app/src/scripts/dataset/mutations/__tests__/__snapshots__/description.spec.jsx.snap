// Vitest Snapshot v1

exports[`UpdateDescription mutation > renders with common props 1`] = `
<DocumentFragment>
  Mutation mock
</DocumentFragment>
`;
